File: Fundamentals_of_Software_Design/2_Algorithms/Пирамидальная сортировка/main.c
#include <stdio.h>
#include <stdlib.h>

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void heapify(int* arr, int n, int i) {
    int largest = i; // Инициализация корня как самого большого
    int left = 2 * i + 1; // Левый дочерний узел
    int right = 2 * i + 2; // Правый дочерний узел

    // Если левый дочерний элемент больше корня
    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    // Если правый дочерний элемент больше самого большого
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    // Если самый большой элемент не корень
    if (largest != i) {
        swap(&arr[i], &arr[largest]);

        // Рекурсивно делаем кучу
        heapify(arr, n, largest);
    }
}

void heap_sort(int* arr, int n) {
    // Строим кучу (max-heap)
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // Извлекаем элементы из кучи по одному
    for (int i = n - 1; i > 0; i--) {
        swap(&arr[0], &arr[i]); // Перемещаем текущий корень в конец
        heapify(arr, i, 0); // Восстанавливаем кучу
    }
}

void print_array(int* arr, int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int arr1[] = {10, 20, 30, 40, 50};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    heap_sort(arr1, n1);
    print_array(arr1, n1); // Вывод: 10 20 30 40 50

    int arr2[] = {5};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    heap_sort(arr2, n2);
    print_array(arr2, n2); // Вывод: 5

    int arr3[] = {3, 1, 4, 1, 5, 9, 2, 6};
    int n3 = sizeof(arr3) / sizeof(arr3[0]);
    heap_sort(arr3, n3);
    print_array(arr3, n3); // Вывод: 1 1 2 3 4 5 6 9

    return 0;
}

File: Fundamentals_of_Software_Design/2_Algorithms/Пирамидальная сортировка/main.py
def heapify(arr, n, i):
    """Приводит поддерево с корнем в i к виду кучи (max-heap)."""
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Меняем местами
        heapify(arr, n, largest)

def heap_sort(arr):
    """Реализует пирамидальную сортировку."""
    n = len(arr)

    # Строим кучу (max-heap)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Извлекаем элементы из кучи по одному
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Меняем корень с последним элементом
        heapify(arr, i, 0)  # Восстанавливаем кучу для уменьшенного массива

# Тестирование
def test():
    # Краевые случаи
    arr1 = []
    heap_sort(arr1)
    assert arr1 == []

    arr2 = [5]
    heap_sort(arr2)
    assert arr2 == [5]

    arr3 = [3, 1, 4, 1, 5, 9, 2, 6]
    expected3 = sorted(arr3)
    heap_sort(arr3)
    assert arr3 == expected3

    arr4 = [10, 20, 30, 40, 50]
    expected4 = [10, 20, 30, 40, 50]
    heap_sort(arr4)
    assert arr4 == expected4

    arr5 = [5, 5, 5, 5, 5]
    expected5 = [5, 5, 5, 5, 5]
    heap_sort(arr5)
    assert arr5 == expected5

    # Тест случайных массивов
    for _ in range(10):
        arr = [random.randint(0, 100) for _ in range(random.randint(0, 20))]
        expected = sorted(arr)
        heap_sort(arr)
        assert arr == expected

    print("Все тесты пройдены!")

test()
